qcodes/instrument/meta/meta_parameter.py

```diff
@@ -0,0 +1,174 @@
+        self._setter = setter
+        self.unit = unit if unit is not None else ''
+        endpoint_names = endpoint_names or [_e.name for _e in endpoints]
+        self._group = None
+
+        # Generate sub parameters if >1 endpoints,
+        # e.g. my_meta_param.X, my_meta_param.Y
+        if len(endpoints) > 1:
+            self._parameters = self._sub_parameters(endpoints, endpoint_names)
+        self._set = self.set
+        self.set = self._set_override
+
+    @property
+    def group(self) -> Optional['MetaGroup']:
+        """
+        The group that this parameter belongs to.
+        """
+        return self._group
+
+    def _sub_parameters(self, endpoints: Tuple[Parameter], endpoint_names: Tuple[str]):
+        """Generate sub parameters if there are > 1 endpoints."""
+        parameters = []
+        for endpoint, endpoint_name in zip(
+            endpoints,
+            endpoint_names
+        ):
+            parameter = self.add_parameter(
+                cls=self.__class__,
+                name=endpoint_name,
+                endpoint=endpoint
+            )
+            parameters.append(parameter)
+        return tuple(parameters)
+
+    def add_parameter(
+        self,
+        cls,
+        name: str,
+        endpoint: Parameter
+    ) -> Parameter:
+        """Add a sub parameter to the meta parameter
+
+        Args:
+            name: Sub parameter name
+            endpoint: Endpoint parameter object
+
+        Returns:
+            Parameter: Parameter instance
+        """
+        assert not hasattr(
+            self, name
+        ), f"Duplicate parameter name {name}."
+        parameter = cls(
+            name=name,
+            endpoints=(endpoint,),
+            endpoint_names=(endpoint.name,),
+            instrument=None
+        )
+        setattr(self, name, parameter)
+        return parameter
+
+    @property
+    def endpoints(self):
+        """Get endpoint parameters"""
+        return self._endpoints
+
+    @property
+    def parameters(self):
+        """Get all sub parameters"""
+        return self._parameters
+
+    def get_raw(self):
+        """Get parameter raw value"""
+        if self.group is not None:
+            return self.group.get_parameters()
+        return self.endpoints[0]()
+
+    def set_raw(self, value: Union[float, Dict[str, ParamDataType]]):
+        """Set parameter raw value
+
+        Args:
+            value: Parameter value to set
+
+        Returns:
+            float: Returns the parameter value
+        """
+        if self._setter is not None:
+            return self._setter(value)
+        elif self.group is not None:
+            self.group.set_parameters(value)
+        else:
+            return self._endpoints[0](value)
+    
+    def _set_override(self, value: ParamDataType = None, **kwargs):
+        """Overide set method such that values can optionally be set using only kwargs"""
+        if value is None:
+            self._set(kwargs)
```

Comment:
Using kwargs without expanding them looks odd and is not general something that I would expect to see but perhaps it makes sense here?

Reply:
Here it's mean to support:
`my_group_param(foo="bar", spam="eggs")`
which then calls
`my_group._set_from_dict({"foo": "bar", "spam": "eggs"})`